src/main/java/org/spongepowered/api/command/args/GenericArguments.java

```diff
@@ -312,40 +320,35 @@ public void parse(CommandSource source, CommandArgs args, CommandContext context
             }
         }
 
+        @Nullable
         @Override
         protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
             return null;
         }
 
         @Override
         public List<String> complete(CommandSource src, CommandArgs args, CommandContext context) {
-            for (Iterator<CommandElement> it = this.elements.iterator(); it.hasNext(); ) {
-                CommandElement element = it.next();
-                Object startState = args.getState();
+            Set<String> completions = Sets.newHashSet();
+            for (CommandElement element : elements) {
+                Object start = args.getState();
                 try {
                     element.parse(src, args, context);
-                    Object endState = args.getState();
+                    if (start.equals(args.getState())) {
+                        completions.addAll(element.complete(src, args, context));
+                    } else if (!completions.isEmpty()) {
+                        completions.clear();
+                    }
                     if (!args.hasNext()) {
-                        args.setState(startState);
-                        List<String> inputs = element.complete(src, args, context);
-                        args.previous();
-                        if (!inputs.contains(args.next())) { // Tabcomplete returns results to complete the last word in an argument.
-                            // If the last word is one of the completions, the command is most likely complete
-                            return inputs;
```

Comment:
Don't remove this behaviour.

Reply:
This is similar to the behavior fixed in #1660. Let's presume we are tab completing a `User` and that users by the names of `dualspiral` and `dual` exists. With this system, entering `dual` and completing will end up completing the *next* element, not to `dualspiral` as I might expect. I removed this because it makes an assumption that the argument is complete rather than letting the user decide that for themselves.

Reply:
This then breaks the behaviour of all custom elements that don't use pattern matching and potentially use this behaviour. We should not break expected behaviour.

Reply:
I'm not sure if this really is expected behavior though, and this change shouldn't break anything. If a user tries to tab complete an argument, I don't think it should try to complete the next one instead.

For example, with a command `/test <user> <item>`, this is what currently happens when tab completing:
* `/test dual` -> All items
* `/test dual (space)` -> All items

This looks nice, but is an issue if the sender was trying to tab complete `dual` to `dualspiral` in the first one; not complete all the items. Instead, I think that this is the expected behavior:
* `/test dual` -> `dual, dualspiral`
* `/test dual (space)` -> All items

Reply:
OK, fair enough, I see what you're getting at now. The space should be a necessary delimiter.

#1660 will supersede part of this. I need to go back to it, test it and merge it, it was built when debugging something for blood then I got busy with other things. If I get to that before this gets sorted, just bear in mind that fix will need to be removed.